### ISSUE TITLE
[SR-12687] Let me add package dependencies that use CMake

### ISSUE THREAD

**Issue by swift-ci** [apple/swift-package-manager#4555]

**Comment by sergeiromanov** [comment]
This is much needed for our project too. We have a cmake c++ project which swift code depends on. Would be great if we could just add cmake project as a dependency to a swift target.

**Comment by neonichu** [comment]
I think doing this would be extremely problematic for build correctness and performance. By definition, SwiftPM would not be able to introspect what is happening in a CMake build, leading to poor parallelism and making it impossible to cache build artifacts in any meaningful way.

**Comment by sergeiromanov** [comment]
The only way to have cmake in xcode is to build .xcodeproj from it which I did. Is it possible to include it into SwiftPM?

**Comment by dschaefer2** [comment]
Absolutely. That is high on my list of things to look at once we get further along in the transition to Swift Build. So many great libraries out there that use CMake that Swift Packages could leverage.

**Comment by furby-tm** [comment]
I’d like to add that this would be monumental for Swift’s ecosystem, as @StingKo mentioned. However, I also noticed some discussion around Xcode-related tooling and I believe it’s crucial to ensure Linux is considered in this effort as well.

**Comment by dschaefer2** [comment]
And Windows ;)

**Comment by dabrahams** [comment]
> I think doing this would be extremely problematic for build correctness and performance. By definition, SwiftPM would not be able to introspect what is happening in a CMake build, leading to poor parallelism and making it impossible to cache build artifacts in any meaningful way.

As usual, I disagree with the SPM philosophy that if you can't control everything, it's not worth doing.  Many people need or want to do this, and they are happy to accept the performance hits. They shouldn't have to roll their own solutions.  We can write a build tool plugin to do this but it won't be nearly as clean.

**Comment by dschaefer2** [comment]
That comment was from a previous regime :). As I mentioned integrating with CMake will be key to SwiftPM's success as we try to grow the ecosystem.

**Comment by dabrahams** [comment]
That's great to hear. And I take it back about the plug-ins. There doesn't appear to be a way to make a plug-in generate a library that will be linked with.

**Comment by dschaefer2** [comment]
Yes, that would be the big part of getting this to work, allow plugins to add more to the build graph, like the libraries that would be produced by the CMake build that can then be depended upon by Swift modules or even other C libraries.

**Comment by jakepetroules** [comment]
> > I think doing this would be extremely problematic for build correctness and performance. By definition, SwiftPM would not be able to introspect what is happening in a CMake build, leading to poor parallelism and making it impossible to cache build artifacts in any meaningful way.
> 
> As usual, I disagree with the SPM philosophy that if you can't control everything, it's not worth doing. Many people need or want to do this, and they are happy to accept the performance hits. They shouldn't have to roll their own solutions. We can write a build tool plugin to do this but it won't be nearly as clean.

@dabrahams The concerns @neonichu has pointed out are still valid of course, but our thinking on this has evolved over time in favor of taking a more balanced approach providing more flexibility to support more use cases.

Agree with @dschaefer2 that some form of CMake integration would be a great addition. Perhaps we could do that through CMake's [package dependency system](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html) (they even have a registry concept) as way of standardizing how Swift packages and CMake projects talk to each other, which could help maintain some degree of structure.

**Comment by OdNairy** [comment]
Could the CMake's [file-api](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html) be useful here?
Or is it already too far from the SPM approach?

**Comment by dschaefer2** [comment]
@OdNairy potentially. The trick will be to pull out the build artifacts that we want to plug into the SwiftPM build graph and do so quickly. Alternatively, the user could provide that in the package manifest but that's more error prone.

**Comment by fbarbat** [comment]
Is there any estimated timeline for this? Is there anything I can do to help? :)

> This is much needed for our project too. We have a cmake c++ project which swift code depends on. Would be great if we could just add cmake project as a dependency to a swift target.

We have the same requirements. Hey @sergeiromanov, how does your build set up look like? Are you creating an XCFramework from .xcodeproj and then wrapping it with an SPM binaryTarget? How do you make sure that the XCFramework and corresponding SPM package is up-to-date with the latest C++ code changes? I'd like the CMake SPM package to be a private package to modularize code, I don't want it to be published, i.e., I don't want to do explicit releases and I just want to reference it from other internal Swift packages in the repo, and rebuild only what's needed when Swift or C++ (CMake side) code changes.

I know how to set this up by having a manual prebuild step for generating an XCFramework for the CMake part and the using a binaryTarget pointing to it but I'd like to avoid that manual prebuild step. Any current workarounds for this? Ideally, I want to press CMD+B and have Xcode/SPM build everything as needed.

Thanks!

**Comment by dschaefer2** [comment]
> Is there any estimated timeline for this? Is there anything I can do to help? :)

This is going to be a pretty major change that will require additions to the Package Manifest and capabilities of build tool plugins to be added to plug everything into resulting build graph. And we want to complete the transition to SwiftBuild so we aren't having to do this twice (i.e. also with the legacy native build system). I'm hoping to start putting ideas together over the summer. Just lots of SwiftBuild work that needs to get done first.

**Comment by sergeiromanov** [comment]
> We have the same requirements. Hey [@sergeiromanov](https://github.com/sergeiromanov), how does your build set up look like? Are you creating an XCFramework from .xcodeproj and then wrapping it with an SPM binaryTarget? How do you make sure that the XCFramework and corresponding SPM package is up-to-date with the latest C++ code changes? I'd like the CMake SPM package to be a private package to modularize code, I don't want it to be published, i.e., I don't want to do explicit releases and I just want to reference it from other internal Swift packages in the repo, and rebuild only what's needed when Swift or C++ (CMake side) code changes.
> 
> I know how to set this up by having a manual prebuild step for generating an XCFramework for the CMake part and the using a binaryTarget pointing to it but I'd like to avoid that manual prebuild step. Any current workarounds for this? Ideally, I want to press CMD+B and have Xcode/SPM build everything as needed.
> 
> Thanks!

@fbarbat First I have a cmake target which produces a dozen of static libs. Then I have another cmake custom target which creates a framework for each of the libs `xcodebuild -create-xcframework -library ...`. These frameworks are then referenced in the swft package. With the right dependencies configured my build process has two steps: 1 - build the custom target; 2 - build the app which imports the swift package.

There is a lot of boiler plate cmake/spm configuration which I'd love to get rid of.

**Comment by ethindp** [comment]
Any update on this? Although I have a problem of wanting the reverse integration: being able to call SPM from CMake somehow. I use dependencies primarily from vcpkg (though if we could get SPM to be able to integrate with that that would pretty much make me an immediate convert to swift build...). The feature this issue wants would just be the cherry on top at that point.

**Comment by dschaefer2** [comment]
I am looking at how SwiftPM can integrate artifacts from a CMake driven build into its build graph. This would make it easier to leverage all the great C/C++ libraries out there.

It's probably worth creating a separate issue for vcpkg integration. I don't have any experience with it.

**Comment by ethindp** [comment]
@dschaefer2 vcpkg is in theory pretty simple. It can be used through cmake (which is the "traditional" way) or, in lieu of that, you just have some root where it's installed in (say) an environment variable, as an example. Then all of your libraries are under `$VCPKG_ROOT/installed/target_triplet/{include|lib}` for release builds and `$VCPKG_ROOT/installed/target_triplet/debug/{include|lib}` for debug builds. Dynamically linked libraries are put under `$VCPKG_ROOT/installed/target_triplet/bin|debug/bin`. vcpkg also works with pkg-config, although I haven't really gotten that to work reliable.

**Comment by dschaefer2** [comment]
I thought the topic sounded familiar. https://github.com/swiftlang/swift-package-manager/issues/4381

I was thinking more of some way to specify in your package manifest that you have a dependency on a port from vcpkg and with the help of a build plugin, SwiftPM can get it built and integrated into the package build graph.

**Comment by ethindp** [comment]
Oooh now that would be superb and would pretty much solve the cmake problem completely. Particularly since if we expand it to cover arbitrary port overlays (which vcpkg already supports) or just custom vcpkg arguments, you could probably then make it possible to link to any C library in existence. In theory anyway.

**Comment by dschaefer2** [comment]
Agreed. I'm pretty fascinated by this. I'll see if I can work it into the changes I want to make.
